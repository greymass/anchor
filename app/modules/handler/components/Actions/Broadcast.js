// @flow
import React, { Component } from 'react';
import { withTranslation } from 'react-i18next';
import { Button, Header, Icon, Label } from 'semantic-ui-react';

class PromptActionBroadcast extends Component<Props> {
  render() {
    const {
      disabled,
      loading,
      onClick,
      t,
      wallet,
    } = this.props;
    return (
      <Button
        animated="fade"
        as="div"
        disabled={disabled}
        floated="right"
        labelPosition="left"
        onClick={onClick}
      >
        <Label
          as="a"
          basic
          pointing="right"
        >
          <Header textAlign="left">
            <Header.Content>
              <Header.Subheader style={{ fontWeight: 'bold' }}>
                {t('handler_action_broadcast_header')}
              </Header.Subheader>
              {wallet.account}@{wallet.authorization}
            </Header.Content>
          </Header>
        </Label>
        <Button
          color="purple"
          icon
          loading={loading}
          style={{ position: 'relative' }}
        >
          <Button.Content hidden style={{ marginTop: '-0.75em' }}>
            <Icon
              name="cloud upload"
              size="large"
              style={{ marginRight: 0 }}
            />
          </Button.Content>
          <Button.Content visible style={{ margin: '0 1em' }}>
            <Icon
              name="cloud"
              size="large"
              style={{ marginRight: 0 }}
            />
          </Button.Content>
        </Button>
      </Button>
    );
  }
}

export default withTranslation('handler')(PromptActionBroadcast);
