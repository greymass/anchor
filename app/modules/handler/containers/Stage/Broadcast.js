// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withTranslation } from 'react-i18next';
import compose from 'lodash/fp/compose';
import { Divider, Form, Grid, Header } from 'semantic-ui-react';

import PromptFragmentTransactionAction from '../../components/Fragment/Transaction/Action';

class PromptStageReview extends Component<Props> {
  render() {
    const {
      blockchain,
      prompt,
      t,
    } = this.props;
    const {
      signed,
      resolved,
    } = prompt;
    const {
      transaction
    } = resolved;
    const { signatures } = signed;
    return (
      <Grid>
        <Grid.Column width={6}>
          <Header>
            {t('handler_containers_stage_broadcast_header')}
            <Header.Subheader>
              {t('handler_containers_stage_broadcast_subheader')}
            </Header.Subheader>
          </Header>
          <Divider />
          <Form>
            <Form.Field>
              <label>API Node</label>
              <Form.Input
                value={blockchain.node}
              />
            </Form.Field>
            <Form.Field>
              <label>{t('handler_containers_stage_broadcast_form_label')}</label>
              <Form.TextArea
                rows={4}
                value={signatures.join('\n')}
              />
            </Form.Field>
          </Form>
        </Grid.Column>
        <Grid.Column width={10}>
          <Header>
            {t('handler_containers_stage_broadcast_grid_header')}
            <Header.Subheader>
              {t('handler_containers_stage_broadcast_grid_subheader')}
            </Header.Subheader>
          </Header>
          {transaction.actions.map((action, index) => (
            <PromptFragmentTransactionAction
              action={action}
              index={index}
              total={transaction.actions.length}
            />
          ))}
        </Grid.Column>
      </Grid>
    );
  }
}

function mapStateToProps(state) {
  return {
    prompt: state.prompt,
  };
}

export default compose(
  withTranslation('handler'),
  connect(mapStateToProps)
)(PromptStageReview);
