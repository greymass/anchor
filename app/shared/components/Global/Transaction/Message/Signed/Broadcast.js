// @flow
import React, { Component } from 'react';
import { withTranslation } from 'react-i18next';
import { Button, Container, Divider, Header, Segment } from 'semantic-ui-react';

import GlobalTransactionViewActions from '../../../Transaction/View/Actions';
import GlobalTransactionViewDetail from '../../../Transaction/View/Detail';
import GlobalTransactionViewFull from '../../../Transaction/View/Full';

export class GlobalTransactionMessageSignedBroadcast extends Component<Props> {
  broadcast = () => {
    const {
      actionname,
      actions,
      transaction
    } = this.props;
    const {
      broadcastTransaction
    } = actions;
    broadcastTransaction(transaction, actionname);
  };
  render() {
    const {
      onClose,
      t,
      transaction
    } = this.props;

    let actions = [];
    if (transaction && transaction.transaction) {
      ({ actions } = transaction.transaction.transaction);
    }
    const signed = (transaction.transaction.signatures.length > 0);
    // console.log(this)
    return (
      <Segment basic>
        {(transaction)
          ? (
            <React.Fragment>
              <Header>
                {t('broadcast_transaction_confirm_details_title')}
                <Header.Subheader>
                  {t('broadcast_transaction_confirm_details_content')}
                </Header.Subheader>
              </Header>
              <GlobalTransactionViewDetail
                broadcastable={signed}
                signed={signed}
                transaction={{ data: transaction }}
              />
              <Container textAlign="center">
                <Button
                  color="purple"
                  content={t('broadcast_transaction_broadcast')}
                  icon="wifi"
                  onClick={this.broadcast}
                  size="large"
                />
                <Button
                  color="grey"
                  content={t('close')}
                  icon="x"
                  onClick={onClose}
                  size="large"
                />
              </Container>
              <Segment>
                <GlobalTransactionViewActions
                  actions={actions}
                />
              </Segment>
            </React.Fragment>
          )
          : (
            <Header>
              {t('broadcast_transaction_no_details_title')}
              <Header.Subheader>
                {t('broadcast_transaction_no_details_content')}
              </Header.Subheader>
            </Header>
          )
        }
      </Segment>
    );
  }
}

export default withTranslation('transaction')(GlobalTransactionMessageSignedBroadcast);
