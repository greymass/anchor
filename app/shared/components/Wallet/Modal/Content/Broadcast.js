// @flow
import React, { Component } from 'react';
import { withTranslation } from 'react-i18next';
import {
  Button,
  Container,
  Divider,
  Form,
  Header,
  Message,
  Segment,
  Tab
} from 'semantic-ui-react';

import GlobalTransactionViewActions from '../../../Global/Transaction/View/Actions';
import GlobalTransactionViewDetail from '../../../Global/Transaction/View/Detail';
import GlobalTransactionViewFull from '../../../Global/Transaction/View/Full';

const { ipcRenderer } = require('electron');

class WalletModalContentBroadcast extends Component<Props> {
  constructor(props) {
    super(props);

    this.state = {
      importedESR: undefined,
      importedTransaction: false
    };
  }

  importFile = () => {
    const { settings } = this.props;
    ipcRenderer.send('openFile', settings.lastFilePath);
    ipcRenderer.once('openFileData', (event, data) =>
      this.props.actions.setTransaction(data));
  }

  broadcast = () => {
    const {
      actions,
      transaction
    } = this.props;
    const {
      broadcastTransaction
    } = actions;
    broadcastTransaction(transaction.data);
  }

  onInputChange = (e, { value }) => {
    this.setState({
      importedESR: value
    });
  }

  onTextAreaChange = (e, { value }) => {
    this.setState({
      importedTransaction: value
    });
  }

  handleImport = () => {
    const {
      actions
    } = this.props;
    const {
      setTransaction
    } = actions;
    const {
      importedTransaction
    } = this.state;

    if (importedTransaction) {
      setTransaction(importedTransaction);
    }
  }

  handleESR = () => {
    const {
      importedESR
    } = this.state;
    if (!importedESR) return false;
    if (importedESR.startsWith('esr:') || importedESR.startsWith('eosio:')) {
      ipcRenderer.send('openUri', importedESR);
    } else {
      ipcRenderer.send('openUri', `esr:${importedESR}`);
    }
  }

  render() {
    const {
      onClose,
      t,
      transaction
    } = this.props;
    const {
      data,
      decoded,
      signed
    } = transaction;
    const {
      importedESR
    } = this.state;
    let actions = [];
    if (data && data.transaction) {
      ({ actions } = data.transaction.transaction);
    }
    if (decoded) {
      ({ actions } = decoded);
    }
    const broadcastable = (signed);
    return (
      <Segment basic>
        {(data)
          ? (
            <React.Fragment>
              <Header>
                {t('broadcast_transaction_confirm_details_title')}
                <Header.Subheader>
                  {t('broadcast_transaction_confirm_details_content')}
                </Header.Subheader>
              </Header>
              <GlobalTransactionViewDetail
                broadcastable={broadcastable}
                signed={signed}
                transaction={transaction}
              />
              <Container textAlign="center">
                {(!signed)
                  ? (
                    <Message error>
                      {t('broadcast_transaction_not_signed')}
                    </Message>
                  )
                  : false
                }
                <Button
                  color="purple"
                  content={t('broadcast_transaction_broadcast')}
                  disabled={!signed}
                  icon="wifi"
                  onClick={this.broadcast}
                  size="large"
                />
                <Button
                  color="grey"
                  content={t('close')}
                  icon="x"
                  onClick={onClose}
                  size="large"
                />
              </Container>
              <Segment>
                <GlobalTransactionViewActions
                  actions={actions}
                />
                <Divider />
                <GlobalTransactionViewFull
                  transaction={decoded || data}
                />
              </Segment>
            </React.Fragment>
          )
          : (
            <Segment basic textAlign="center">
              {(transaction.error) && (
                <Message
                  content={t('broadcast_transaction_invalid_error')}
                  error
                />
              )}
              <Tab
                defaultActiveIndex={0}
                renderActiveOnly={false}
                panes={[
                  {
                    menuItem: {
                      key: 'import_file',
                      icon: 'upload',
                      content: t('broadcast_transaction_import_file_title')
                    },
                    pane: {
                      key: 'import_file',
                      content: (
                        <Segment basic>
                          <Header>
                            <Header.Subheader>
                              {t('broadcast_transaction_import_file_content')}
                            </Header.Subheader>
                          </Header>
                          <Button
                            color="green"
                            content={t('broadcast_transaction_import_file')}
                            icon="upload"
                            onClick={this.importFile}
                            size="large"
                          />
                        </Segment>
                      )
                    }
                  },
                  {
                    menuItem: {
                      key: 'paste',
                      icon: 'paste',
                      content: t('broadcast_transaction_paste_json_title')
                    },
                    pane: {
                      key: 'paste',
                      content: (
                        <Form>
                          <Form.TextArea
                            onChange={this.onTextAreaChange}
                            placeholder={t('broadcast_transaction_import_json_label')}
                          />
                          <Button
                            color="blue"
                            content={t('broadcast_transaction_import_json')}
                            icon="paste"
                            onClick={this.handleImport}
                            size="large"
                          />
                        </Form>
                      )
                    }
                  },
                  {
                    menuItem: {
                      key: 'esr',
                      icon: 'paste',
                      content: t('broadcast_transaction_paste_esr_title')
                    },
                    pane: {
                      key: 'esr',
                      content: (
                        <Form>
                          <Form.Input
                            onChange={this.onInputChange}
                            placeholder={t('broadcast_transaction_import_esr_label')}
                            value={importedESR}
                          />
                          <Button
                            color="blue"
                            content={t('broadcast_transaction_import_esr')}
                            icon="paste"
                            onClick={this.handleESR}
                            size="large"
                          />
                        </Form>
                      )
                    }
                  }
                ]}
              />
              <Message
                content={t('broadcast_transaction_note')}
                icon="info circle"
                warning
              />
            </Segment>
          )
        }
      </Segment>
    );
  }
}

export default withTranslation('transaction')(WalletModalContentBroadcast);
