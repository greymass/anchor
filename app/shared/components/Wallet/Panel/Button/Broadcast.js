// @flow
import React, { Component } from 'react';
import { withTranslation } from 'react-i18next';
import { Icon, Menu } from 'semantic-ui-react';

import GlobalTransactionModal from '../../../Global/Transaction/Modal';
import WalletModalContentBroadcast from '../../Modal/Content/Broadcast';

class WalletPanelButtonBroadcast extends Component<Props> {
  render() {
    const {
      actions,
      blockExplorers,
      color,
      settings,
      system,
      t,
      transaction
    } = this.props;
    // Prevent render in cold wallet
    if (settings.walletMode === 'cold') return false;
    return (
      <GlobalTransactionModal
        actionName="TRANSACTION_BROADCAST"
        actions={actions}
        blockExplorers={blockExplorers}
        content={(
          <WalletModalContentBroadcast
            actions={actions}
            settings={settings}
          />
        )}
        customTrigger={(
          <Menu.Item
            as="a"
            style={{ color }}
          >
            <Icon name="wifi" size="large" />
            {(!settings.sidebarCollapsed)
              ? <p>{t('wallet_panel_button_broadcast_transaction')}</p>
              : false
            }
          </Menu.Item>
        )}
        icon="wifi"
        title={t('wallet_panel_wallet_broadcast')}
        settings={settings}
        system={system}
        transaction={transaction}
      />
    );
  }
}

export default withTranslation('wallet')(WalletPanelButtonBroadcast);
