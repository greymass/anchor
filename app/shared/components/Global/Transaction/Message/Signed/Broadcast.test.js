import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GlobalTransactionMessageSignedBroadcast } from './Broadcast';

const mockProps = {
  onClose: () => null,
  t: text => text,
  transaction: {
    transaction: {
      signatures: [],
      transaction: {
        actions: []
      }
    }
  },
};

describe('GlobalTransactionMessageSignedBroadcast', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<GlobalTransactionMessageSignedBroadcast {...mockProps} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
