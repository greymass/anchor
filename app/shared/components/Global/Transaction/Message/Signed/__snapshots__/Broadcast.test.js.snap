// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalTransactionMessageSignedBroadcast renders correctly 1`] = `
<Segment
  basic={true}
>
  <Header>
    broadcast_transaction_confirm_details_title
    <HeaderSubheader>
      broadcast_transaction_confirm_details_content
    </HeaderSubheader>
  </Header>
  <Translate(GlobalTransactionViewDetail)
    broadcastable={false}
    signed={false}
    transaction={
      Object {
        "data": Object {
          "transaction": Object {
            "signatures": Array [],
            "transaction": Object {
              "actions": Array [],
            },
          },
        },
      }
    }
  />
  <Container
    textAlign="center"
  >
    <Button
      as="button"
      color="purple"
      content="broadcast_transaction_broadcast"
      icon="wifi"
      onClick={[Function]}
      size="large"
    />
    <Button
      as="button"
      color="grey"
      content="close"
      icon="x"
      onClick={[Function]}
      size="large"
    />
  </Container>
  <Segment>
    <Translate(GlobalTransactionViewActions)
      actions={Array []}
    />
    <Divider />
    <Translate(GlobalTransactionViewFull)
      transaction={
        Object {
          "transaction": Object {
            "signatures": Array [],
            "transaction": Object {
              "actions": Array [],
            },
          },
        }
      }
    />
  </Segment>
</Segment>
`;
